handle axe violations issues resolved

mohanqxf2 opened the pull request:
- Updated the snapshot_util.py to handle the resolved, updated axe accessibility violations and displayed a better message for users.

mohanqxf2 commented:
I have tested the following cases:

1. New violation added:
When the snapshot finds  new violation item [i.e dictionary_item_added] in the snapshot it says as New violation added
`{ "description":  "id":   "impact":   "nodes":  "tags":   }` 

---
2. Violation resolved:
When the snapshot has resolved the existing violation item [i.e dictionary_item_removed] in the snapshot it says as  Violation resolved
`{ "description":  "id":   "impact":   "nodes":  "tags":   }` 

---
3. Violation Node updated:
When the snapshot violations are changed/updated/modified [i.e values_changed] in the snapshot, it says as Violation updated along with the Old value and New value of the violations


---
4. Violation Node added:
When the snapshot violations Nodes/Related nodes found newly [i.e iterable_item_added] in the snapshot,, It says as Violation Node added along with the node details

_Remove the Node item in the existing snapshot to test this case._
```
{
                        "data": null,
                        "id": "aria-label",
                        "impact": "serious",
                        "message": "aria-label attribute does not exist or is empty",
                        "relatedNodes": []
}
```


---
5. Violation Node resolved
When the snapshot violations Nodes/Related nodes resolved[i.e iterable_item_removed] in the snapshot,, It says as Violation Node resolved with the node details

_Add the Node item in the existing snapshot to test this case._

avinash010 commented:
@mohanqxf2 If the violation gets removed shouldn't the test status be Passed with a message to the user to update the snapshot?

avinash010 commented:
@mohanqxf2 The tests that you performed aren't complete. You can try other conditions too like
* What if you add some nodes and remove others for the same root element?
*  Add some nodes and remove others for different root elements.
* Similarly add new violations and remove others.

I could find some issues with these tests

mohanqxf2 commented:
> @mohanqxf2 If the violation gets removed shouldn't the test status be Passed with a message to the user to update the snapshot?

For this case, need to slightly change the logic part, as of now the snapshots are comparing with each other, if any changes found then the test will be failed, if matches then the test will pass

I will update the login to handle if any violations/nodes resvoled, then the test should be pass[for this case, the current and existing snapshot will not match because of violation will be resolved in the current snapshot]

mohanqxf2 commented:
> @mohanqxf2 If the violation gets removed shouldn't the test status be Passed with a message to the user to update the snapshot?

@avinash010  I have updated the snapshot_util.py to allow the tests to pass when _violation dict or violation node_ gets resolved and also logs the resolved violation details on the output and also get updates the one violation text file.
- When _Violation node_ gets resolved
![image](https://github.com/user-attachments/assets/859c39e9-941a-4097-a630-6ded239e9222)

- When _Violation dict_ gets resolved
![image](https://github.com/user-attachments/assets/e839977d-1bf7-4e78-ad7b-b09213000f21)

avinash010 reviewed (commented):
@mohanqxf2 Overall the functionality works fine. However what if there were no violations before?
I guess we added a logic to display message saying no existing violations were found creating violation file. I guess that looks broken

avinash010 commented:
@mohanqxf2 Can you just come up with a flow chart of how things would flow here?
I can see some issues here. Coming up with a flowchart will help you how to design your flow and test better
Also how did you test for no violation path?

avinash010 commented on utils/snapshot_util.py:
```diff
@@ -88,6 +88,44 @@ def initialize_snapshot(self, snapshot_dir, page, current_violations=None):
+                for violation in current_violations
+            ]
+            self.log_violations_to_file(new_violations, log_path)
+            return False, new_violations
```
Wouldn't this bea  duplicate way to handle same condition

avinash010 commented:
@mohanqxf2 This looks nice. Now i guess you have testing scenarios too before your make any new changes to this file
![Uploading Axe_Violation.jpeg…]()

mohanqxf2 commented:
@avinash010 as per your suggestions saving the new violations found to the existing snapshot file, also i have updated the flowchart as per this new case as well.

mohanqxf2 commented:
@avinash010 I have made the following updations to the script,
a. removed the functionality for saving only new-found violations to an existing snapshot
b. Implemented the --snapshot-update cmd line arg to update the existing snapshots with a copy of current_violations
c. Updated the logic when current_violations is None:
  i. Current violations is None and existing violations is None-> No violations found -> test pass
  ii. Current violations is None and existing violations is Not None-> All violations are resolved -> test pass
